File: Chapter 02/Python/2.4.1 Constraints In Action/export_controller.py
from database import Database

class ExportController:
    def __init__(self):
        self.db = Database()

    # Simulating a Flask / FastAPI endpoint
    def export_user_data(self, user_id):
        try:
            user_data = self.db.fetch_user_data(user_id)

            if not user_data:
                print("  [HTTP 404] User not found.")
                return

            # Simple CSV conversion
            headers = "id,name,email\n"
            csv_row = f"{user_data['id']},{user_data['name']},{user_data['email']}\n"
            csv_data = headers + csv_row

            # Simulating HTTP Response
            print("  [HTTP 200] OK")
            print("  [Headers] Content-Type: text/csv")
            print(f"  [Headers] Content-Disposition: attachment; filename=\"user_data_{user_id}.csv\"")
            print("\n--- File Body ---")
            print(csv_data, end="")
            print("-----------------")

        except Exception as e:
            print(f"  [HTTP 500] Export failed: {e}")

File: Chapter 02/C#/2.4.1 Constraints In Action/ExportController.cs
using System;
using System.Threading.Tasks;

namespace Chapter02.ConstraintsInAction;

public class ExportController
{
    private readonly Database _db = new();

    // Simulating an ASP.NET Endpoint
    public async Task ExportUserDataAsync(string userId)
    {
        try
        {
            var userData = await _db.FetchUserDataAsync(userId);

            if (userData == null)
            {
                Console.WriteLine("  [HTTP 404] User not found.");
                return;
            }

            // Simple CSV conversion
            var headers = "id,name,email\n";
            var csvRow = $"{userData.Id},{userData.Name},{userData.Email}\n";
            var csvData = headers + csvRow;

            // Simulating HTTP Response
            Console.WriteLine("  [HTTP 200] OK");
            Console.WriteLine("  [Headers] Content-Type: text/csv");
            Console.WriteLine($"  [Headers] Content-Disposition: attachment; filename=\"user_data_{userId}.csv\"");
            Console.WriteLine("\n--- File Body ---");
            Console.Write(csvData);
            Console.WriteLine("-----------------");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"  [HTTP 500] Export failed: {ex.Message}");
        }
    }
}